### DISCUSSION TITLE
Instant Tab Sync between Computers

### DISCUSSION THREAD

**Discussion by xanthos84** [zen-browser/desktop#2400] [Ideas]
Hi everyone,

I love the idea of zen browser, and i can see how far this can go in some years!
I am one of the "ARC switchers" and I am missing my instant sync feature of all open tabs to my browser between all my computers.
If you switch from one device to another, its so handy to have your complete "session" with all open tabs and workspaces ready to work on.

And as there are no other browsers around with that feature beside ARC, it would be a huge USP for Zen in my opinion.

**Comment by Hary309** [comment]
Yeah, this is very weird arc was the first to come up with this idea.

**Comment by Gamedirection** [comment]
Tab Session Manager has been a good work around for myself.

**Reply by Gamedirection** [reply]
https://floccus.org/ has also been a great solution if you utilize Nextcloud at all too.

**Reply by nms808** [reply]
> Tab Session Manager has been a good work around for myself.

This!

**Reply by rgdcastro** [reply]
> Tab Session Manager has been a good work around for myself.

This doesn't work if you have folders.

**Comment by friendywill** [comment]
*Probably naive solution*

From what I understand, if this feature was implemented, the pain point of workspaces not syncing to Firefox mobile (I understand that the synchronised tab feature exists, but the workspace organisation of tabs, pinned tabs and favourites does not) still exists.

So, why not have a reserved bookmark folder for the Zen Browser? Here is what I am thinking:

- Prefix a top-level bookmark folder with `zen:`
- Suffix this bookmark folder with `${Profile}`
- Each folder contains a `${workspace}` folder, and their respective tabs.

Zen would then detect, and deal with this accordingly, then hiding the bookmarks in the bookmark bar prefixed with `zen:` (I am unsure if this is problematic). This would require a diff like system between the synchronised bookmarks and the state of the workspaces, at least, in theory.

This would alleviate, and mitigate some pain points with the sync feature suggestion:

- Their would be, at the very least, another way to access your pinned tabs on mobile. (Synchronised tabs seems to work temperamentally for me)
- You will only have to sign into Mozilla to have these particular items synchronised.

I am not sure if modifying user bookmarks is touching on a sensitive topic for some, so obviously something like this would have to be an optional setting.

Other potential advantages:

- Settings synchronisation, where another tab named after a reserved string could be used. That tab will contain some URL, eg `https://zen-browser.app/users?=${USERID}`. Once the tab is recgonized by Zen, it will open that URL, which will have Mozilla SSO authentication to protect the contents. No config file to worry about.
- Future support for folders and sub-folders, where workspaces could have tab folders and tab sub-folders and it would translate well into the bookmark. Currently, subfolders in the bookmarks (representing workspaces) could just be handled by ignoring sub-folders but including their respective tabs in the workspace.
- Future support for trees, where subfolders could just have empty names.

I am not a full-time developer, so if someone could please tell me if this would be difficult to implement, because at the surface it seems like an easy option.

**Reply by FancyFurret** [reply]
I would love if this was implemented

**Reply by qdawgao-dawei** [reply]
> _Probably naive solution_
> 
> From what I understand, if this feature was implemented, the pain point of workspaces not syncing to Firefox mobile (I understand that the synchronised tab feature exists, but the workspace organisation of tabs, pinned tabs and favourites does not) still exists.
> 
> So, why not have a reserved bookmark folder for the Zen Browser? Here is what I am thinking:
> 
> * Prefix a top-level bookmark folder with `zen:`
> * Suffix this bookmark folder with `${Profile}`
> * Each folder contains a `${workspace}` folder, and their respective tabs.
> 
> Zen would then detect, and deal with this accordingly, then hiding the bookmarks in the bookmark bar prefixed with `zen:` (I am unsure if this is problematic). This would require a diff like system between the synchronised bookmarks and the state of the workspaces, at least, in theory.
> 
> This would alleviate, and mitigate some pain points with the sync feature suggestion:
> 
> * Their would be, at the very least, another way to access your pinned tabs on mobile. (Synchronised tabs seems to work temperamentally for me)
> * You will only have to sign into Mozilla to have these particular items synchronised.
> 
> I am not sure if modifying user bookmarks is touching on a sensitive topic for some, so obviously something like this would have to be an optional setting.
> 
> Other potential advantages:
> 
> * Settings synchronisation, where another tab named after a reserved string could be used. That tab will contain some URL, eg `https://zen-browser.app/users?=${USERID}`. Once the tab is recgonized by Zen, it will open that URL, which will have Mozilla SSO authentication to protect the contents. No config file to worry about.
> * Future support for folders and sub-folders, where workspaces could have tab folders and tab sub-folders and it would translate well into the bookmark. Currently, subfolders in the bookmarks (representing workspaces) could just be handled by ignoring sub-folders but including their respective tabs in the workspace.
> * Future support for trees, where subfolders could just have empty names.
> 
> I am not a full-time developer, so if someone could please tell me if this would be difficult to implement, because at the surface it seems like an easy option.

It's exactly the same as my idea. I think this is a suitable solution and hope it can be realized.

**Comment by Alintya** [comment]
Currently my biggest pain point as an arc switcher myself (well, besides firefox/gecko having countless issues due to everything being made for chromium).

**Reply by yashcrest** [reply]
sadly im still on arc :(

**Comment by the-ult** [comment]
Yeah, same here.
Really would like to switch to Zen. But having too many issues.
- Sites not working properly
- settings/short-cuts sometimes working and the next day not anymore
- 1password not being unlocked (although zen browser is added to the allowed browsers in 1password)
   - although this might be a 1password problem..

But the main thing, might be the (missing) of the auto-sync of my open tabs. Loving that about Arc

**Reply by yashcrest** [reply]
this and the little arc feature.

**Comment by tobias-hay** [comment]
Bumping this! Missing my tabs and sessions syncing between my mac and pc :(

**Comment by barrymichaeldoyle** [comment]
Another week another bump :) It would be an absolute game changer if I can have instant (or at least near instant syncing between my Mac and PC!

**Comment by SCGilbert** [comment]
Bumping as well - I'd love to make the swap to Zen and this is currently the last hurdle for me to overcome. I'm currently using Chrome's experimental "Tab Groups Save and Sync V2" and "Tab Groups Save UI Update" to do this within chrome/brave/vivaldi and it works ~sort of~alright. Iirc Arc and Edge are the only browsers available that treated this as a first-class feature, but neither will support Linux/Windows/MacOS together which is (unfortunately) essential for my work.

**Reply by SCGilbert** [reply]
Coming back to this as I've just found that Edge does actually support Linux now (previously you had to use older versions that were giving me trouble) so that's at least an option now for people in a similar situation as me. Only downside is that now you're on Edge 😏

Zen, pls save me.

**Comment by mFontoura** [comment]
Just found out this was not a feature after coming home from work and not having my tabs that I organized at work :/

**Comment by amitbartfeld** [comment]
Bumping this too - it's one of the most important features for me since I constantly switch between my laptop and home PC and I need sync. Until now it was one of the only features I've enjoyed in ARC's Windows version. I've used Chrome's Tab group sync feature which gets the job done but I really want to fully move to Zen.
Hope this feature will be added soon!

**Comment by smilingpeanut** [comment]
Bumping as well. I thought that Firefox Sync did this so I was excited to try out Zen, but now that I have it set up on both computers, there is no sync and I has a sad.

**Comment by omsharp** [comment]
Bump! 

I just installed Zen and finished organizing my tabs on my desktop, then went to install it on my laptop and found it doesn't sync tabs!! This is a huge deal breaker for me. I was making the move from Arc but without this feature I don't think I can make it.

**Comment by jallalcode** [comment]
Bumping. I feel like this is really significant to user experience

**Comment by 12th-devs** [comment]
+1

**Comment by phendan** [comment]
This is the single biggest feature that's currently keeping me stuck on Arc as someone who goes between devices a lot, where manually copying over a bunch of tabs all the time would be a nightmare. As much as Zen's superior performance and active development make switching very tempting, having the browser feel like just a window to my work wherever I happen to be has been incredibly useful and hard to let go of.

I also understand that this is probably one of the more complex features to implement though. It interacts with a lot of other features and there are a bunch of (presumably) non-obvious decisions to be made (around where to sync to and reconciling different sets of tabs). So basically I'm expecting to have to be patient here, but this is currently **_the_** feature that would make Zen my daily driver if implemented.

**Comment by JxJxxJxJ** [comment]
I use Edge because of the workspaces sync between computers. I have 3 devices at home and they're all syncronized. If Zen had just that single feature I'd 100% switch no doubt

**Comment by TeoPXL** [comment]
I'd have to agree, this is maybe the BIGGEST thing I miss in zen (Windows & Mac user)

**Comment by hyperdotbat** [comment]
bump, I thought of switching from vanilla Firefox to Zen, and damn so sad we don't have a nice workspace with open tabs sync, would ease my switch from Windows to Linux

**Comment by aquafraternal** [comment]
I'm yet another Arc refugee.  I absolutely love Zen.  The only two things missing for me are the arc-like syncing of tabs/workspaces and an iPhone version.  I can get by without the iPhone version, but syncing is an awfully big one that's keeping me tied to the sinking ship that is Arc.  I'm using Zen wherever I can, but I'm crossing my fingers for this feature to be added.

**Comment by waLLxAck** [comment]
Bump! From an Arc user.

**Comment by AntoineArt** [comment]
Adding my bump, I cannot switch to zen because of that

**Comment by kennytrbl** [comment]
Bumping this as a former Arc user. I'd also like my Zen Mods to be synced.

**Comment by tribute2ro** [comment]
Bump. I love Zen and miss this from Arc.

**Comment by Ac-nl** [comment]
A must-have feature imho

**Comment by davidmickelsen** [comment]
Giving this another bump. I love Zen, fantastic browser; I would love to be able to sync my tabs and workspaces between devices.

**Comment by adelphi23** [comment]
Bump! It is my main pain point as I have to switch between Mac and Linux computers at least three times in a regular day. It is just frustrating to keep consistency without a decent automated sync feature.

**Comment by UndercoverNerd5** [comment]
+1 I think we are all in need of this and it becomes a matter of time until the devs add this in

**Comment by freisenhauer** [comment]
Also +1. Currently using Arc to sync between mac/windows .. Now i replace my windows with linux and Arc is not supported by linux, so Zen would gain much more attraction if this feature was available.

**Comment by heyJordanParker** [comment]
I'm all in for this. Definitely the biggest weakness of Zen right now.

**Comment by tbreslein** [comment]
Another +1. It's quite a hassle to manage my open tabs through bookmarks. It works, sorta, but it's a pain 😅

**Comment by 74796a** [comment]
So having just wasted an hour of my life let me see if I've got this straight? Someone working on Zen decided it would be a good idea to arbitrarily force ALL of their users to conform to a random aesthetic preference (arguably an objective regression from a UX design perspective which i have a degree in) by breaking part of the interface without removing it? You just randomly changed the way opening a tab/window has worked for as long as the internet and browsers have existed and foisted your precious little preference on every last one of your end users without so much as announcing it or providing an opt-out for people who just want it to work the way it's always worked?  

This has to be the single most arrogant, short-sighted, self-important decision by a software company since Apple decided they knew better than everyone else and removed the back button.  In a way it's a good thing because that single lost hour of my time has prevented me from making the mistake of investing further time or resources supporting this project. It does more than that though, it gives me a few new names (Zen principals who would have been aware of this change) to add to my list of people whose involvement in any project of any type is a huge red flag and probably a basis not to contribute in and of itself.  Just stunningly, breathtakingly cringe and customer-hostile behavior.  Like whoever's idea this was is just a low quality person in general and anyone who went along with it is a mediocre pushover at best; it's just incredibly childish and repulsive behavior by a developer in any context.

**Comment by Aguiar16** [comment]
+1, this is **the dealbreaker** for me to go full Zen as my main browser :(
I work on various projects and three different machines. If I can't always have the open tabs on whatever PC I'm on, it feels like my brain just melted, and trying to remember what I was doing is frustrating.

edit: typo

**Comment by tanaybhomia** [comment]
They do it by making users create a Arc account which even led to that Security issues, So I dont think that feature is going to come.

**Reply by friendywill** [reply]
Yes and connecting to the internet leads to security concerns, why would this security concern lead to not implementing a highly requested feature? Especially when basic security measures could have avoided the issue Arc had?
If people are concerned, make it a toggle with it off by default.
Also, people can more easily catch such defects considering that Zen is open source.

**Comment by jamie9090** [comment]
I've switched from Arc to Edge for now, but would love to use Zen. 

Edge Workspaces is the best I've found so far, and it goes beyond Arc because it also currently works to sync tabs between Android and Mac with some limitations (no tablet support yet).

Would be interested in a Chrome or Firefox implementation if something similar comes.

**Comment by martin-braun** [comment]
Personally, for now I would be happy to have it sync between my macOS and Linux, both are installed on the same MacBook, both have Syncthing with my phone in the middle. So I don't need any hot sync, but an export/import to a file (automatically) would help me a lot. This is rather attracting to me, because there is no third party involved, but it is a bit of a hassle and not really a solution for the masses. I will share my progress, because it is a mandatory step for me, before I can even begin using Linux as a daily driver.

**Reply by Garfounkel** [reply]
Thanks for the initiative, please share updates. Also using Syncthing, that would be a suitable solution for me too.

**Reply by martin-braun** [reply]
@Garfounkel I will, but I am very slow, because I am still on macOS working on things that pay the bills. If I come up with something, I will share ofc. :)

**Reply by martin-braun** [reply]
@Garfounkel I started my journey in the attempt to write an `.stignore` for SyncThing. The idea is to move the `zen` folder in your sync folder while adding the `.stignore`, which is a negative ignore list (white lising files instead of blacklisting files). You could then symlink it back and Zen would work just fine, which it does.

I found the `about:config` stuff and extensions + their settings. ZenMods can't be synced, because they contain absolute paths. I found spaces (containers), but neither tab containers nor (pinned) tabs. I guess this stuff is serialized into one of the many sqlite DBs, which makes it a bit tricky to do reliable sync of these, because they are opened and updated during the browser process lifetime and contain possibly more information that is not suitable to sync.

I guess a much better approach might be to build an import/export shell script and run it during sesssion start/stop. It rules hot syncing out, but could work for when you are on one frame at a time, i.e. home pc vs working pc, etc.

If anyone knows in which of the many sqlite files tabs and tab containers are stored, please let me know.

**Reply by martin-braun** [reply]
@Garfounkel [took a bit xd](https://github.com/zen-browser/desktop/discussions/2400#discussioncomment-18200682)

**Comment by kameqdev** [comment]
Bumping as a former Arc user. It's one of the features I am missing in Zen the most

**Comment by Zeroble** [comment]
I really want this, too....

**Comment by FutureKen** [comment]
This should be the most wanted and needed feature, let's do this!!

**Comment by zyhe16** [comment]
I use a dual-boot laptop with NixOS and Windows, and having this feature would greatly simplify my workflow and make switching between different computers much easier. I truly believe most users would appreciate this addition as much as I would. Also, are there any updates on this feature?

**Comment by alexandrehaas** [comment]
I consider this as an essential key feature to implement before stable release !

**Comment by staus** [comment]
This is such a bummer that there still isn't an approach to this. 
I think FriendyWill's [suggestion](https://github.com/zen-browser/desktop/discussions/2400#discussioncomment-11633871) sounds quite good actually. At first it sounds like a hack, but really it's just using an existing functionality and adding a UI around it.

**Comment by nirbhaykwatra** [comment]
Bumping this. The lack of tab syncing across workspaces and computers is single handedly preventing me from switching to Zen. I use my laptop at work and my desktop at home so switching machines is an everyday occurrence for me. Going through the pain of copying over all my tabs between my laptop and PC is simply a bridge too far. I absolutely love Zen and if/when this feature is introduced, I am never looking back. Zen Team, please consider implementing this feature.

**Comment by Ruszero01** [comment]
Zen's interaction, appearance, and Glance make me really like it, but Firefox's various issues are also a headache. Using this browser makes me overcome many difficulties. I used crx installer to solve some crx extension problems, and UA spoofing to solve some web compatibility issues. For example, some website feature switches can only be displayed in Chrome, but after spoofing the UA, they can surprisingly work normally in Firefox. It's just that the websites simply haven't adapted, and the low number of Firefox users leads to many websites not bothering to adapt.

**Comment by sklknn** [comment]
+1

**Comment by atqamz** [comment]
+1

**Comment by tiimonn** [comment]
+1 :c

**Comment by joshjhall** [comment]
+1... this is the last major piece I liked in Arc still missing in Zen

**Comment by KarlJussila** [comment]
+1 from a long time Arc user who really doesn't want to have to keep using _Microsoft Edge_ of all things in a Linux/Windows dual boot and multi-device setup. I'm fine with the limitations of Firefox, but I just can't live without automatic tab syncing.

**Comment by SamuelYancy** [comment]
Bump. Tap syncing across devices is THE feature I require to use Zen over Arc.

**Comment by abelsuzihan** [comment]
+1... this is a crucial feature, hope Zen could support it

**Comment by friedrichjaum** [comment]
+1

**Comment by Marabbekka** [comment]
+1

**Comment by ochronus** [comment]
+1

**Comment by Dufery** [comment]
+1

**Comment by lemonilemon** [comment]
Bump! Loving Zen as my Arc alternative, but this bit’s still missing.

**Comment by ocueye2** [comment]
The problem with this is that there needs to be back end infrastructure that stores your tabs and recalls it. that would also mean you would need an account and would also change the whole privacy focused idea. you could do a LAN peer to peer version using qr codes  where the qr code encodes IP, port, and secret but if a device gets disconnected and both change local ip's then you will have to reconnect them. you could also just store tab data in a qr code but it could be to big for that.

**Reply by fredrikaverpil** [reply]
A simple and somewhat ok solution to that is to provide the ability to specify a filepath where this data would be saved?
On macOS it could go to iCloud. On Windows, OneDrive. Or DropBox, Google Drive etc etc. That would however not necessarily solve cross-OS syncing (iCloud isn't a thing on Linux for example).

**Reply by heyJordanParker** [reply]
We have bookmarks that are already synced. No extra backend is necessary. Just the correct parsing & a hidden bookmarks folder.

(as one solution, obviously; more are certainly available)

**Reply by ocueye2** [reply]
> A simple and somewhat ok solution to that is to provide the ability to specify a filepath where this data would be saved? On macOS it could go to iCloud. On Windows, OneDrive. Or DropBox, Google Drive etc etc. That would however not necessarily solve cross-OS syncing (iCloud isn't a thing on Linux for example).

could just do git as a private repo

**Reply by nirbhaykwatra** [reply]
Obsidian has something similar, you can have an auto-pull set up with a private repo. There could also be the possibility of a self-hosted solution for sync.

**Reply by cliff-4** [reply]
Git repo, Self hosting, Google Drive, Drop Box, all valid solutions to the storage problem.

**Comment by aquafraternal** [comment]
Syncthing does a pretty good job of providing secure peer-to-peer file syncing between computers, even over the Internet. I’m not going to claim to know the ins-and-outs of that software package, so maybe there are pitfalls I’m not aware of, but it might provide reasonable solution to the info storage issue.

**Comment by liamssi** [comment]
+1 , have been one of the early adopters of zen and this is my biggest complain so far , hope this get implemented in the  near future .

**Comment by geraldzm** [comment]
+1

**Comment by Stormers16** [comment]
+1 Please implement this and I will full transition to Zen. I work in 3 different computer, having an auto-sync tabs is essencial for me.

**Comment by luweiCN** [comment]
+1

**Comment by bdbch** [comment]
I think this is the last thing (besides stability) that I miss from Arc.

I tried to read myself a bit into Firefox's sync possibilities and I have two questions:

1. Does Zen want to go for Firefox Sync as its solution? As far as I know the sync server can be self hosted so it would be a nice option to default to Mozilla's default server but give the option to use a custom server in the UI (currently you can only set the URL via about:config)
2. If it'd go for Firefox Sync, is it actually possible? As far as I know Firefox uses Sync Engines to sync content between devices - so in theory we could create a ZenWorkspace sync engine that would take over syncing data between two devices if I understand correctly?

**Reply by heyJordanParker** [reply]
With "Simple yet powerful" as a core value – the 3rd party sync is not ideal. Firefox Sync will likely be the way to go.
(right now _window_ sync for workspaces is in development and that's step 1)

**Comment by bujoriosif** [comment]
+1

**Comment by roland-KA** [comment]
Adding myself to this list. Sync of tabs/tab folders is a core feature I need, before I can leave Arc.

And I don't understand, why Issue #8211 has been closed!

**Comment by juzi214032** [comment]
+1

**Comment by bdbch** [comment]
I am currently working on this on a fork of mine, will provide a PR when I think it's ready. I'm very close - what works:

- Sync Pinned Tabs via Firefox Sync Engine
- Sync Tab Folders via Firefox Sync Engine
- Sync Essential Tabs via Firefox Sync Engine
- Wait for the correct Workspaces before sync
- Use correct tab containers for tabs
- Sync back icon changes on folders

What's not working yet:
- Syncing back the updated tab name, looking into this right now

**Reply by juzi214032** [reply]
Before merging the pull request, could a test version be provided for use?

**Reply by bdbch** [reply]
That's something cheff will need to decide, but I think there's no workflow for that.

We're waiting for the window sync PR to be merged anyways.

**Reply by Davarpe** [reply]
Reading this made me happy! This is what I needed in Zen too! :D

**Reply by Stormers16** [reply]
Does someone has a guide on how pull his version to my Zen? If the official Zen version gets updated, will it update mine?
Sorry if the questions are dumb, its been a long time since I last used any code.😭

**Reply by briansunter** [reply]
If you land this I'll come back to zen. I just really need the sync across desktop and laptop. Great work!

**Reply by bdbch** [reply]
@Stormers16 you could try to build it yourself from my branch directly.
https://github.com/bdbch/zen-desktop/tree/feat/tab-sync

**Reply by francisqureshi** [reply]
Exciting! Thank you for your work on this! :)

**Comment by AlbertHuangT** [comment]
+1

**Comment by leafmoes** [comment]
+1

**Comment by lioryasur** [comment]
This is super important for me as well, I migrated to Zen on my laptop but am still using Arc for my desktop because it's a hassle to set everything up with my sidebar favorites and stuff all over again. And even when I will there will be no continued sync.

**Comment by dmp-git** [comment]
This is the last feature I really really really miss and is pushing me back from switching from Arc.
Hoping the branch made by @bdbch will be merged soon

**Comment by idan-walton** [comment]
+1

**Comment by franklegolasyoung** [comment]
+1. PLSSSSSSSS

**Comment by twentytwokhz** [comment]
+1 It's unusual to have *everything else* sync, but the tabs.. 🙄

**Comment by LGTiscar** [comment]
Bumping!

**Comment by mmailaender** [comment]
Please stop spamming this thread with +1 entries or everything else which does not support the solution of this issue. A lot of people are subscribed to this issue and it's quite annoying to get constantly notified only because someone added a +1 comment. If you like this idea, upvote the idea at the top, but please stop spamming.

**Reply by bst1n** [reply]
It's important to understand that this feature has been eagerly awaited, and that it's hard to wait for it when you've just come from Arc. It's really the crucial thing we've been missing. It makes you want to comment again and again...

**Reply by bujoriosif** [reply]
> It's important to understand that this feature has been eagerly awaited, and that it's hard to wait for it when you've just come from Arc. It's really the crucial thing we've been missing. It makes you want to comment again and again...

Agree 200%

**Reply by rmchale** [reply]
Also the lack of update since the original post seems a bit of an issue. Is it done? Does it need work? Where does this new work land on the roadmap?

**Reply by mr-cheffy** [reply]
> Where does this new work land on the roadmap?

Now that cross-window sync is merged, this should become a possibility really soon!

**Reply by myrrhkhan** [reply]
@mr-cheffy awesome! My apologies for asking a pretty loaded question, but what's the difference between implementing window syncing on the same device versus different devices? I'm curious to see how the implementations differ and what additional challenges come with device syncing
Thanks for your work!

**Reply by mr-cheffy** [reply]
The main thing for cross-device is that you would need a server 😅

**Reply by tobias-hay** [reply]
> The main thing for cross-device is that you would need a server 😅

Is cross-device next on the roadmap or is that a while out yet? I can understand that a server might be a blocker

**Reply by mr-cheffy** [reply]
Yep, its on our todo list

**Reply by tobias-hay** [reply]
> Yep, its on our todo list

Awesome can't wait. One of the last big things im missing.

**Reply by SilverEzhik** [reply]
> The main thing for cross-device is that you would need a server 😅

Is it something that cannot be handled by Firefox Account infrastructure?

**Reply by IsaiahHolman** [reply]
> The main thing for cross-device is that you would need a server 😅

Forgive my ignorance if I'm completely wrong, but isn't a peer-to-peer sync a viable alternative to a server? 

In my mind, this wouldn't be a perfectly convenient solution as devices would need to be on the same network to sync (e.g. I couldn't expect my work laptop tabs to sync to my personal laptop), but there are so many other benefits:
- Avoids the cost of a server. 
- No server maintenance required.
- No server downtime - sync feature always working. Even when Zen stops being updated.
- If encrypted, it is more secure. No concern about data breaches.
- Private. No concerns that Zen or server maintainers are looking at data.

**Reply by mr-cheffy** [reply]
> Is it something that cannot be handled by Firefox Account infrastructure?

Yep, there's a PR for that in the works. I wasn't so knowledgeable about FF's sync when I wrote the comment.

> Forgive my ignorance if I'm completely wrong, but isn't a peer-to-peer sync a viable alternative to a server?

Its certainly an option, but not many people are that technical so I think a more mainstreamed solution, such as using the existing firefox sync, would be more ideal

**Reply by nirbhaykwatra** [reply]
Is there any chance for a self-hosted solution? Maybe as a second option to Firefox Sync?

**Reply by SilverEzhik** [reply]
> Is there any chance for a self-hosted solution? Maybe as a second option to Firefox Sync?

Firefox Sync is already self-hostable, there's a Docker image that works pretty well here: https://github.com/dan-r/syncstorage-rs-docker

**Comment by quantumfate** [comment]
Zen stores pinned tabs and spaces inside a SQL light database, right? the flake https://github.com/0xc000022070/zen-browser-flake to configure zen-browser declarative supports the declaration of pins and workspaces. Under the hood it's just executing SQL statements on the internal SQL light database ([see here](https://github.com/0xc000022070/zen-browser-flake/blob/8098fd20ce52c78d0a6227ab675f29f1195642d3/hm-module.nix#L434)). Perhaps this is a good starting point as well as a reference to implement something that's already working somewhere else?

One of the limitations this solution had was that Zen needed to be closed or else you'd risk corrupting the internal SQLlite database. But if this is implemented internally with appropriate singleton locks, I could see this to be a good starting point. However, I wouldn't know how to start implementing this because I have never actually developed web-browsers before.

Another thing I believe it's worth taking a look at is how Firefox itself is able to send tabs to other devices and even keeps a complete index of what tabs are opened where. Zen certainly adds more complexity with limiting bookmarks to certain workspaces (which is something that breaks if you sync bookmarks through firexfox sync) hence drawing a flow chart with all the moving parts would definitely be helpful.

I loved Zen so much when I was still using nix-os because I could easily reproduce my entire environment within zen on other machines with a single command. But since I want to be on Arch the zen-browser is sadly not an option at the moment as I can't easily reproduce my workflow on a new machine.

I wouldn't mind collaborating on this tho because I desperately need this feature - it's an absolute hell to work in any browser besides zen.

**Comment by naxey** [comment]
Now that #13984 is merged and #14470 is closed, is cross-device sync for normal tabs still planned?
I mean actual sidebar/session sync, not Firefox View or same-device Window Sync.

#12531 included this, but #13984 ended up syncing only pinned and Essential tabs. Was normal-tab sync deferred or dropped? Is there another issue to follow?

**Comment by martin-braun** [comment]
Just dropping by to share my self-tested Zen import/export function that allows me to sync my entire Zen through Syncthing between macOS and Linux. It syncs workspaces, tabs/folders, it also syncs all extensions and extension settings. Been using it for a while between Artix (Flatpak --user) and MacOS (Brew --cask), successfully.

It does not sync the normal browser settings, because things like keymaps bite each other on different OSes.

```sh
# zen profile import/export
zep() {
	if [ $# -lt 1 ] || [ "$1" = "-h" ] || [ "$1" = "--help" ]; then
		echo "Usage: zep [-h, --help] [-e, --export, -i, --import] file" >&2
		echo "Imports or exports all Zen browser profiles, excluding storage" >&2
		echo "and other unimportant or incompatible files." >&2
		echo "For Linux, assumes usage of Flatpak (user-scoped)." >&2
		echo >&2
		return 129
	fi
	if [ -z "$2" ]; then
		echo "File argument missing." >&2
		return 1
	fi
	if [ -z "$ZEN_WORKING_DIR" ]; then
		echo "ZEN_WORKING_DIR not set." >&2
		return 9
	fi
	if [ "$OSFAM" = "Darwin" ]; then
		if pgrep -f "$ZEN_PROC" > /dev/null 2>&1; then
			echo "Zen Browser is still running. Please close it first." >&2
			return 2
		fi
	elif [ "$OSFAM" = "GNU_Linux" ]; then
		if pgrep -x "$ZEN_PROC" > /dev/null 2>&1; then
			echo "Zen Browser is still running. Please close it first." >&2
			return 2
		fi
	else
		echo "Unsupported OS: $OSFAM" >&2
		return 1
	fi
	for p in "$ZEN_WORKING_DIR"/Profiles/*; do
		[ -d "$p" ] || continue
		grep '^user_pref("zen.keyboard.shortcuts.version",' "$p/prefs.js" \
			| tail -n 1 \
			| sed 's/.*,[[:space:]]*\([0-9][0-9]*\));/\1/' \
			> "$p/zen-keyboard-shortcuts.version"
	done
	if [ "$1" = "-e" ] || [ "$1" = "--export" ]; then
		shift
		cat > "$ZEN_WORKING_DIR/zen.exc" << EOF
*/._*
*/.DS_Store
.DocumentRevisions-V100
.fseventsd
./.TemporaryItems
/.Trashes/
Profiles/*/*backup*
Profiles/*/*Cache*
Profiles/*/*crash*
Profiles/*/*Crash*
Profiles/*/*dump*
Profiles/*/*logs*
Profiles/*/*report*
Profiles/*/*ShutdownDuration*
Profiles/*/*backup.json
Profiles/*/*cache*
Profiles/*/*datareporting*
Profiles/*/*download*
Profiles/*/*jetpack*
Profiles/*/*saved-telemetry-pings*
Profiles/*/*thumbnails*
Profiles/*/.parentlock
Profiles/*/.startup-incomplete
Profiles/*/AlternateServices*
Profiles/*/Partitions/**
Profiles/*/Photos
Profiles/*/QuotaManager
Profiles/*/RunningChromeVersion
Profiles/*/Telemetry.ShutdownTime.txt
Profiles/*/bookmarkbackups
Profiles/*/cert-revocations/*.stash
Profiles/*/compatibility.ini
Profiles/*/cur
Profiles/*/gcontactsync_log*.txt
Profiles/*/msgFilterRules.dat
Profiles/*/panacea.dat
Profiles/*/photos
Profiles/*/safebrowsing-failedupdate
Profiles/*/serviceworker.txt
Profiles/*/storage
Profiles/*/tmp
Profiles/*/weave
Profiles/*/zen-keyboard-shortcuts.json
Profiles/*/zen-keyboard-shortcuts.version
EOF
		printf "Exporting '%s' to file '%s'…\n" "$ZEN_WORKING_DIR" "$1"
		tar -C "$ZEN_WORKING_DIR" -X "$ZEN_WORKING_DIR/zen.exc" \
			--no-xattrs -cf "$1" Profiles installs.ini profiles.ini && \
		(
			cd "$ZEN_WORKING_DIR" \
				&& find Profiles -type d -path \
					'Profiles/*/storage/default/moz-extension+++*' -prune -print
		) | tar -C "$ZEN_WORKING_DIR" --no-xattrs --exclude='*/._*' -rf "$1" -T -
	elif [ "$1" = "-i" ] || [ "$1" = "--import" ]; then
		shift
		printf "Importing file '%s' into '%s'…\n" "$1" "$ZEN_WORKING_DIR"
		tar --overwrite -xf "$1" -C "$ZEN_WORKING_DIR" Profiles installs.ini profiles.ini 2> /dev/null \
			|| tar -xf "$1" -C "$ZEN_WORKING_DIR" Profiles installs.ini profiles.ini \
			|| return $?
		for p in "$ZEN_WORKING_DIR"/Profiles/*; do
			[ -s "$p/zen-keyboard-shortcuts.version" ] || continue
			printf 'user_pref("zen.keyboard.shortcuts.version", %s);\n' \
				"$(cat "$p/zen-keyboard-shortcuts.version")" >> "$p/prefs.js"
		done
	fi
}
```

With this, I don't really care about online sync, I want to control my data anyways, so Syncthing it is. Obviously you have to stop Zen before export/import and it's not handy if you have multiple machines with Zen open, but for dual booting scenarios this is perfect.

Note you have to initialize Zen on all OSes and merge the `installs.ini` and `profiles.ini`, so all installs point to the same profile.

Then set your environment variables in your dotfiles where you also put the function above. For example in Linux:

```sh
export OSFAM="$(uname -o | tr ' ' '_' | tr '/' '_' | tr -d '"')"
export ZEN_PROC='zen|zen-bin|zen-browser'
export ZEN_WORKING_DIR="$HOME/.var/app/app.zen_browser.zen/config/zen"
```

or MacOS:

```sh
export OSFAM="$(uname -o | tr ' ' '_' | tr '/' '_' | tr -d '"')"
export ZEN_PROC='/Zen\.app/Contents/MacOS/zen'
export ZEN_WORKING_DIR="$HOME/Library/Application Support/zen"
```

Make backups!

I'm out. :wave: cc @Garfounkel
